File: Stripe/STPEphemeralKeyManager.m
//
//  STPEphemeralKeyManager.m
//  Stripe
//
//  Created by Ben Guo on 5/9/17.
//  Copyright © 2017 Stripe, Inc. All rights reserved.
//

#import "STPEphemeralKeyManager.h"

#import "NSError+Stripe.h"
#import "STPEphemeralKey.h"
#import "STPPromise.h"

static NSTimeInterval const DefaultExpirationInterval = 60;
static NSTimeInterval const MinEagerRefreshInterval = 60*60;

@interface STPEphemeralKeyManager ()
@property (nonatomic) STPEphemeralKey *ephemeralKey;
@property (nonatomic) NSString *apiVersion;
@property (nonatomic) id keyProvider;
@property (nonatomic, readwrite, assign) BOOL performsEagerFetching;
@property (nonatomic) NSDate *lastEagerKeyRefresh;
@property (nonatomic) STPPromise<STPEphemeralKey *>*createKeyPromise;
@end

@implementation STPEphemeralKeyManager

- (instancetype)initWithKeyProvider:(id)keyProvider
                         apiVersion:(NSString *)apiVersion
              performsEagerFetching:(BOOL)performsEagerFetching {
    self = [super init];
    if (self) {
        NSAssert([keyProvider conformsToProtocol:@protocol(STPCustomerEphemeralKeyProvider)] || [keyProvider conformsToProtocol:@protocol(STPIssuingCardEphemeralKeyProvider)], @"Your STPEphemeralKeyProvider must either implement `STPCustomerEphemeralKeyProvider` or `STPIssuingCardEphemeralKeyProvider`.");
        _expirationInterval = DefaultExpirationInterval;
        _keyProvider = keyProvider;
        _apiVersion = apiVersion;
        _performsEagerFetching = performsEagerFetching;
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(handleWillForegroundNotification)
                                                     name:UIApplicationWillEnterForegroundNotification
                                                   object:nil];
    }
    return self;
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self
                                                    name:UIApplicationWillEnterForegroundNotification
                                                  object:nil];
}

- (void)setExpirationInterval:(NSTimeInterval)expirationInterval {
    _expirationInterval = MIN(expirationInterval, 60*60);
}

- (BOOL)currentKeyIsUnexpired {
    return self.ephemeralKey && self.ephemeralKey.expires.timeIntervalSinceNow > self.expirationInterval;
}

- (BOOL)shouldPerformEagerRefresh {
    return self.performsEagerFetching && (!self.lastEagerKeyRefresh || self.lastEagerKeyRefresh.timeIntervalSinceNow > MinEagerRefreshInterval);
}

- (void)handleWillForegroundNotification {
    // To make sure we don't end up hitting the ephemeral keys endpoint on every
    // foreground (e.g. if there's an issue decoding the ephemeral key), throttle
    // eager refreshes to once per hour.
    if (!self.currentKeyIsUnexpired && self.shouldPerformEagerRefresh) {
        self.lastEagerKeyRefresh = [NSDate date];
        [self getOrCreateKey:^(__unused STPEphemeralKey * _Nullable ephemeralKey, __unused NSError * _Nullable error) {
            // getOrCreateKey sets the self.ephemeralKey. Nothing left to do for us here
        }];
    }
}

- (void)_createKey {
    STPJSONResponseCompletionBlock jsonCompletion = ^(NSDictionary *jsonResponse, NSError *error) {
        STPEphemeralKey *key = [STPEphemeralKey decodedObjectFromAPIResponse:jsonResponse];
        if (key) {
            [self.createKeyPromise succeed:key];
        } else {
            // the API request failed
            if (error) {
                [self.createKeyPromise fail:error];
            } else {
                // the ephemeral key could not be decoded
                [self.createKeyPromise fail:[NSError stp_ephemeralKeyDecodingError]];
                if ([self.keyProvider conformsToProtocol:@protocol(STPCustomerEphemeralKeyProvider)]) {
                    NSAssert(NO, @"Could not parse the ephemeral key response following protocol STPCustomerEphemeralKeyProvider. Make sure your backend is sending the unmodified JSON of the ephemeral key to your app. For more info, see https://stripe.com/docs/mobile/ios/standard#prepare-your-api");
                } else if ([self.keyProvider conformsToProtocol:@protocol(STPIssuingCardEphemeralKeyProvider)]) {
                    NSAssert(NO, @"Could not parse the ephemeral key response following protocol STPCustomerEphemeralKeyProvider. Make sure your backend is sending the unmodified JSON of the ephemeral key to your app. For more info, see https://stripe.com/docs/mobile/ios/standard#prepare-your-api");
                }
                NSAssert(NO, @"Could not parse the ephemeral key response. Make sure your backend is sending the unmodified JSON of the ephemeral key to your app. For more info, see https://stripe.com/docs/mobile/ios/standard#prepare-your-api");
            }
        }
        self.createKeyPromise = nil;
    };
    
    if ([self.keyProvider conformsToProtocol:@protocol(STPCustomerEphemeralKeyProvider)]) {
        id<STPCustomerEphemeralKeyProvider> provider = self.keyProvider;
        [provider createCustomerKeyWithAPIVersion:self.apiVersion completion:jsonCompletion];
    } else if ([self.keyProvider conformsToProtocol:@protocol(STPIssuingCardEphemeralKeyProvider)]) {
        id<STPIssuingCardEphemeralKeyProvider> provider = self.keyProvider;
        [provider createIssuingCardKeyWithAPIVersion:self.apiVersion completion:jsonCompletion];
    }
}

- (void)getOrCreateKey:(STPEphemeralKeyCompletionBlock)completion {
    if (self.currentKeyIsUnexpired) {
        completion(self.ephemeralKey, nil);
    } else {
        if (self.createKeyPromise) {
            // coalesce repeated calls into one request
            [[self.createKeyPromise onSuccess:^(STPEphemeralKey *key) {
                completion(key, nil);
            }] onFailure:^(NSError *error) {
                completion(nil, error);
            }];
        } else {
            self.createKeyPromise = [[[STPPromise<STPEphemeralKey *> new] onSuccess:^(STPEphemeralKey *key) {
                self.ephemeralKey = key;
                completion(key, nil);
            }] onFailure:^(NSError *error) {
                completion(nil, error);
            }];
            [self _createKey];
        }
    }
}

@end

File: Stripe/STPEphemeralKeyManager.swift
//
//  STPEphemeralKeyManager.swift
//  Stripe
//
//  Created by Ben Guo on 5/9/17.
//  Copyright © 2017 Stripe, Inc. All rights reserved.
//

import Foundation
import UIKit

typealias STPEphemeralKeyCompletionBlock = (STPEphemeralKey?, Error?) -> Void
class STPEphemeralKeyManager: NSObject {
  /// If the current ephemeral key expires in less than this time interval, a call
  /// to `getOrCreateKey` will request a new key from the manager's key provider.
  /// The maximum allowed value is one hour – higher values will be clamped.

  private var _expirationInterval: TimeInterval = 0.0
  var expirationInterval: TimeInterval {
    get {
      _expirationInterval
    }
    set(expirationInterval) {
      _expirationInterval = TimeInterval(min(expirationInterval, 60 * 60))
    }
  }
  /// If this value is YES, the manager will eagerly refresh its key on app foregrounding.
  private(set) var performsEagerFetching = false

  /// Initializes a new `STPEphemeralKeyManager` with the specified key provider.
  /// - Parameters:
  ///   - keyProvider:               The key provider the manager will use.
  ///   - apiVersion:                The Stripe API version the manager will use.
  ///   - performsEagerFetching:     If the manager should eagerly refresh its key on app foregrounding.
  /// - Returns: the newly-initiated `STPEphemeralKeyManager`.
  @objc init(
    keyProvider: Any?,
    apiVersion: String,
    performsEagerFetching: Bool
  ) {
    super.init()
    assert(
      keyProvider is STPCustomerEphemeralKeyProvider
        || keyProvider is STPIssuingCardEphemeralKeyProvider,
      "Your STPEphemeralKeyProvider must either implement `STPCustomerEphemeralKeyProvider` or `STPIssuingCardEphemeralKeyProvider`."
    )
    expirationInterval = DefaultExpirationInterval
    self.keyProvider = keyProvider
    self.apiVersion = apiVersion
    self.performsEagerFetching = performsEagerFetching
    NotificationCenter.default.addObserver(
      self,
      selector: #selector(handleWillForegroundNotification),
      name: UIApplication.willEnterForegroundNotification,
      object: nil)
  }

  /// If the retriever's stored ephemeral key has not expired, it will be
  /// returned immediately to the given callback. If the stored key is expiring, a
  /// new key will be requested from the key provider, and returned to the callback.
  /// If the retriever is unable to provide an unexpired key, an error will be returned.
  /// - Parameter completion: The callback to be run with the returned key, or an error.
  @objc dynamic func getOrCreateKey(_ completion: @escaping STPEphemeralKeyCompletionBlock) {
    if currentKeyIsUnexpired() {
      completion(ephemeralKey, nil)
    } else {
      if let createKeyPromise = createKeyPromise {
        // coalesce repeated calls into one request
        createKeyPromise.onSuccess({ key in
          completion(key, nil)
        }).onFailure({ error in
          completion(nil, error)
        })
      } else {
        createKeyPromise = STPPromise<STPEphemeralKey>.init().onSuccess({ key in
          self.ephemeralKey = key
          completion(key, nil)
        }).onFailure({ error in
          completion(nil, error)
        })
        _createKey()
      }
    }
  }

  @objc internal var ephemeralKey: STPEphemeralKey?
  private var apiVersion: String?
  private var keyProvider: Any?
  @objc internal var lastEagerKeyRefresh: Date?
  private var createKeyPromise: STPPromise<STPEphemeralKey>?

  deinit {
    NotificationCenter.default.removeObserver(
      self,
      name: UIApplication.willEnterForegroundNotification,
      object: nil)
  }

  func currentKeyIsUnexpired() -> Bool {
    return ephemeralKey != nil
      && (ephemeralKey?.expires.timeIntervalSinceNow ?? 0.0) > expirationInterval
  }

  func shouldPerformEagerRefresh() -> Bool {
    return performsEagerFetching
      && (lastEagerKeyRefresh == nil
        || (lastEagerKeyRefresh?.timeIntervalSinceNow ?? 0.0) > MinEagerRefreshInterval)
  }

  @objc func handleWillForegroundNotification() {
    // To make sure we don't end up hitting the ephemeral keys endpoint on every
    // foreground (e.g. if there's an issue decoding the ephemeral key), throttle
    // eager refreshes to once per hour.
    if !currentKeyIsUnexpired() && shouldPerformEagerRefresh() {
      lastEagerKeyRefresh = Date()
      getOrCreateKey({ _, _ in
        // getOrCreateKey sets the self.ephemeralKey. Nothing left to do for us here
      })
    }
  }

  func _createKey() {
    let jsonCompletion =
      { (jsonResponse: [AnyHashable: Any]?, error: Error?) in
        let key = STPEphemeralKey.decodedObject(fromAPIResponse: jsonResponse)
        if let key = key {
          self.createKeyPromise?.succeed(key)
        } else {
          // the API request failed
          if let error = error {
            self.createKeyPromise?.fail(error)
          } else {
            // the ephemeral key could not be decoded
            self.createKeyPromise?.fail(NSError.stp_ephemeralKeyDecodingError())
            if self.keyProvider is STPCustomerEphemeralKeyProvider {
              assert(
                false,
                "Could not parse the ephemeral key response following protocol STPCustomerEphemeralKeyProvider. Make sure your backend is sending the unmodified JSON of the ephemeral key to your app. For more info, see https://stripe.com/docs/mobile/ios/standard#prepare-your-api"
              )
            } else if self.keyProvider is STPIssuingCardEphemeralKeyProvider {
              assert(
                false,
                "Could not parse the ephemeral key response following protocol STPIssuingCardEphemeralKeyProvider. Make sure your backend is sending the unmodified JSON of the ephemeral key to your app. For more info, see https://stripe.com/docs/mobile/ios/standard#prepare-your-api"
              )
            }
            assert(
              false,
              "Could not parse the ephemeral key response. Make sure your backend is sending the unmodified JSON of the ephemeral key to your app. For more info, see https://stripe.com/docs/mobile/ios/standard#prepare-your-api"
            )
          }
        }
        self.createKeyPromise = nil
      } as STPJSONResponseCompletionBlock

    if keyProvider is STPCustomerEphemeralKeyProvider {
      weak var provider = keyProvider as? STPCustomerEphemeralKeyProvider
      provider?.createCustomerKey(withAPIVersion: apiVersion ?? "", completion: jsonCompletion)
    } else if keyProvider is STPIssuingCardEphemeralKeyProvider {
      weak var provider = keyProvider as? STPIssuingCardEphemeralKeyProvider
      provider?.createIssuingCardKey(withAPIVersion: apiVersion ?? "", completion: jsonCompletion)
    }
  }
}

private let DefaultExpirationInterval: TimeInterval = 60
private let MinEagerRefreshInterval: TimeInterval = 60 * 60
